### ISSUE TITLE
[BUG] Persistent data storage in Python controller bindings is not power loss safe.

### ISSUE THREAD

**Issue by kepstin** [project-chip/connectedhomeip#36654]
### Reproduction steps

A Home Assistant user reported an issue where following a power loss, all of their connected Matter devices had disappeared.

From the user's provided logs: [matter-logs.txt](https://github.com/user-attachments/files/17937239/matter-logs.txt)
It looks like the json file was corrupted (with partially written data), and upon reading the corrupted file the Matter SDK decided to reset and start from a blank storage.

The code responsible for writing the file is here:

https://github.com/project-chip/connectedhomeip/blob/e98f2b239aae657a10043bad9aa61f4717d059be/src/controller/python/chip/storage/__init__.py#L164-L183

The logic here does an in-place overwrite of the file. The problem this causes is that if power is cut at any point after the write is started, but before the fsync(), then the user may be left with a corrupt partially written file which contains a mix of data - parts of the new file and parts of the old file.

I'd recommend the following logic instead:

* (optional) Make a backup of chip.json by copying it to chip.json.backup, leaving the existing file in place. (Run fsync on the newly written file to ensure the backup is updated before the original file gets replaced.)
* Write the new data to a temporary file, like chip.json.new.
* After writing the data, flush and fsync to ensure the file contents are persisted on the disk. This is needed to ensure ordering - make sure that the file contents will be fully written before the file gets renamed.
* Rename the temporary file to the final filename, atomically replacing the existing file.
* Run fsync on the renamed file to ensure the rename is committed to storage.

### Bug prevalence

Rarely - requires unlucky timing of power loss

### GitHub hash of the SDK that was being used

c197bc2db8c2079aa2af78c55fc647d048a70b5b

### Platform

python

### Anything else?

I originally reported this issue on Home Assistant's python-matter-server: https://github.com/home-assistant-libs/python-matter-server/issues/977 but it turns out that the responsible code was actually in the Matter SDK, not provided by Home Assistant.

**Comment by nerijus** [comment]
Someone should probably create a PR.

**Comment by kepstin** [comment]
I would be happy to make a PR, except that I do not want to sign the Matter Working Group CLA.

I have intentionally avoided posting code due to that, although I've provided a list of the changes that would be needed in the PR description, if someone else would like to work on this.

**Comment by nerijus** [comment]
Could you provide the needed changes in `diff -u` or `git diff` format?

**Comment by kepstin** [comment]
No, I cannot provide the changes in such a format (which is essentially the same as making a pull request) because I have not and do not want to sign the CLA.

If i do write the code and post it here without signing the CLA, it may cause legal problems for any contributors who have signed the CLA. (If they copy portions of the code I posted in their submission, they may be in violation of the terms of the CLA or of my copyright of the changed code).

**Comment by kepstin** [comment]
@nerijus that's some large language model generated slop/spam and has nothing to do with this issue.

**Comment by nerijus** [comment]
Yeah, thought so, sorry, deleted.
